inc/telemetry/namespace.php

```diff
@@ -0,0 +1,445 @@
+		] );
+	}
+
+	// Set the group data.
+	$environment = get_environment_details();
+	if ( $environment['id'] ) {
+		Segment::group( [
+			'userId' => $user['id'],
+			'groupId' => $environment['id'],
+			'traits' => $environment['traits'],
+		] );
+	}
+
+	$initialized = true;
+}
+
+/**
+ * Track an event in Segment.io.
+ *
+ * @param array $message The event details.
+ * @return void
+ */
+function track( array $message ) {
+	initialize();
+
+	// Add user ID if missing.
+	if ( ! isset( $message['userId'] ) ) {
+		$user = get_segmentio_user_details();
+		$message['userId'] = $user['id'];
+	}
+
+	Segment::track( $message );
+}
+
+/**
+ * Check user has opted in to tracking.
+ *
+ * @param WP_User|null $user Current user object if available.
+ * @return boolean|null
+ */
+function is_user_opted_in( ?WP_User $user = null ) : ?bool {
+	if ( ! $user ) {
+		$user = wp_get_current_user();
+	}
+	if ( ! $user->exists() ) {
+		return null;
+	}
+
+	$did_opt_in = get_user_meta( $user->ID, META_OPT_IN, true );
+
+	if ( $did_opt_in === '' ) {
+		return null;
+	}
+
+	return intval( $did_opt_in ) === 1;
+}
+
+/**
+ * Get details for the current user.
+ *
+ * @return array
+ */
+function get_segmentio_user_details() : array {
+	$current_user = wp_get_current_user();
+	$auto_id = substr( bin2hex( $current_user->user_registered ), 0, 8 );
+	$did_opt_in = is_user_opted_in( $current_user );
+
+	if ( ! $did_opt_in ) {
+		return [
+			'id' => $auto_id,
+			'opt_in' => false,
+		];
+	}
+
+	$email = $current_user->user_email;
+	$id = bin2hex( $current_user->user_email );
+	if ( Altis\get_environment_type() === 'local' ) {
+		// Create distinct_id (important for Mixpanel).
+		$id = $auto_id;
+		$email = 'no-reply+' . $auto_id . '@altis.dev';
+	}
+
+	$traits = [
+		'first_name' => $current_user->first_name,
+		'last_name' => $current_user->last_name,
+		'email' => $email,
+		'username' => $current_user->user_login,
+		'created' => $current_user->user_registered,
+		'avatar' => get_avatar_url( $current_user->ID ),
+	];
+
+	/**
+	 * Filter user traits delivered to segment.io.
+	 *
+	 * @param array $traits User traits to send to Segment.io.
+	 */
+	$traits = (array) apply_filters( 'altis.telemetry.user_traits', $traits );
+
+	return [
+		'id' => $id,
+		'opt_in' => true,
+		'traits' => $traits,
+	];
+}
+
+/**
+ * Get a unique ID for the current (local) environment.
+ *
+ * Generates a persistent, unique ID for the current local environment.
+ *
+ * @return string|null Null for non-local environments, unique ID otherwise.
+ */
+function get_local_install_id() : ?string {
+	if ( Altis\get_environment_type() !== 'local' ) {
+		return null;
+	}
+
+	$id = get_network_option( null, 'altis_tracking_id', null );
+	if ( ! empty( $id ) ) {
+		return $id;
+	}
+
+	// Hash the registration date of the first site to get a unique ID.
+	$site = get_blog_details( get_main_site_id(), false );
+	$generated = 'local-' . substr( hash( 'sha1', $site->registered ), 0, 6 );
```

Comment:
This might be a bit edge-case, but it's possible that two or more sites have the same registration timestamp, if done programmatically. Why not use something that is actually required to be unique, like the ID full domain? 🤔 (Maybe I don't understand how/what this is used for, or why the date is being used to generate the hash.)

Reply:
These are user IDs across development machines, so ID = 1 and domain is going to overlap across all devs on the same project generally.

Reply:
Aha, I see. You want to distinguish multiple local development environments...?